# **Maestría en Inteligencia Artificial Aplicada**

## Curso: **Procesamiento de Lenguaje Natural**

### Tecnológico de Monterrey

### Prof Luis Eduardo Falcón Morales

### **Actividad en Equipo - Semanas 4 y 5**

### **Vectores Embebidos de HuggingFace**

#### **Nombres y matrículas de los integrantes del equipo:**



- A01769076 | Alonso Pedrero Martínez
- A01797612 | Lilian Margarita Álvarez Iglesias
- A01235844 | Luis Felipe Neri Alvarado Fregoso


In [1]:
# Aqui deberan incluir todas las librerias que requieran durante esta actividad:

import pandas as pd
import numpy as np
import os
import re
from sklearn.model_selection import train_test_split
import string
import nltk
from collections import Counter

os.environ.setdefault('HF_HUB_DISABLE_XET', '1')
os.environ.setdefault('HF_HUB_DISABLE_SYMLINKS_WARNING', '1')

nltk.data.path.insert(0, os.path.abspath(os.path.join('..', '.venv', 'nltk_data')))

En esta actividad deberás utilizar los datos de tres archivos que se encuentran en el repositorio de la UCI llamados **amazon_cells_labelled.txt**, **imdb_labelled.txt** y   **yelp_labelled.txt**. Cada uno de estos archivos corresponden a comentarios de usuarios que adquirieron un celular a través de la plataforma de Amazon, de comentarios que dejaron usuarios sobre palículas y series en la plataforma de IMDb y sobre servicios de comida dejados en la plataforma de Yelp.

La información del problema y de los archivos están basados en el repositorio de la UCI cuya liga es la siguiente:

https://archive.ics.uci.edu/dataset/331/sentiment+labelled+sentences



# **Pregunta - 1:**



Descarga los 3 archivos de la plataforma de la UCI indicado previamente y genera un nuevo DataFrame de Pandas con ellos.

**Llama simplemente "df" a dicho DataFrame.**




In [2]:
PATH = 'sentiment labelled sentences'
if not os.path.exists(PATH):
    PATH = '../files'

df_amazon = pd.read_csv(os.path.join(PATH, 'amazon_cells_labelled.txt'), sep='	', names=['review','label'], header=None, encoding='utf-8')
dfi_imdb = pd.read_csv(os.path.join(PATH, 'imdb_labelled.txt'), delimiter='	', names=['review','label'], header=None, encoding='utf-8')
df_yelp = pd.read_csv(os.path.join(PATH, 'yelp_labelled.txt'), sep='	', names=['review','label'], header=None, encoding='utf-8')

In [3]:
df_amazon.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   review  1000 non-null   str  
 1   label   1000 non-null   int64
dtypes: int64(1), str(1)
memory usage: 15.8 KB


In [4]:
dfi_imdb.info()

<class 'pandas.DataFrame'>
RangeIndex: 748 entries, 0 to 747
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   review  748 non-null    str  
 1   label   748 non-null    int64
dtypes: int64(1), str(1)
memory usage: 11.8 KB


In [5]:
df_yelp.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   review  1000 non-null   str  
 1   label   1000 non-null   int64
dtypes: int64(1), str(1)
memory usage: 15.8 KB


In [6]:
newdfi=[]
dfi_list = dfi_imdb.values.tolist()

for row in dfi_list:

    review = str(row[0])
    label = str(row[1])

    text = review + '\t' + label

    matches = re.findall(r'(.*?)\t([01])(?:\n|$)', text, re.DOTALL)

    for rev, lab in matches:

        rev = rev.strip()

        if rev != '':
            newdfi.append([rev, int(lab)])

# convertir a DataFrame
clean_dfi_imdb = pd.DataFrame(newdfi, columns=['review', 'label'])

In [7]:
clean_dfi_imdb.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   review  1000 non-null   str  
 1   label   1000 non-null   int64
dtypes: int64(1), str(1)
memory usage: 15.8 KB


In [8]:
df = pd.concat([df_amazon, clean_dfi_imdb, df_yelp], ignore_index=True) 
# *********** Aquí termina la sección de agregar código *************


In [9]:
# Verifiquemos la información del DataFrame:

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   review  3000 non-null   str  
 1   label   3000 non-null   int64
dtypes: int64(1), str(1)
memory usage: 47.0 KB


In [10]:
# Y mostremos sus primeros registros:

df.head()

,review,label
0,So there is no way for me to plug it in here i...,0
1,"Good case, Excellent value.",1
2,Great for the jawbone.,1
3,Tied to charger for conversations lasting more...,0
4,The mic is great.,1


# **Pregunta - 2:**

Proceso de limpieza. Aplica el proceso de limpieza que consideres adecuado.











In [11]:
# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


# ************* Inicia la sección de agregar código:*****************************

X = df.review
y = df.label

assert X.shape == (3000,)
assert y.shape == (3000,)  

In [12]:
def clean_tok(doc):
  doc_cleaned = ''.join(
    char for char in doc.lower()
    if char not in string.punctuation and not char.isdigit())
  
  tokens = re.findall(r'[a-z]+', doc_cleaned)
  
  tokens = [
    word for word in tokens
    if len(word) > 1]

  return tokens

In [13]:
Xclean = [clean_tok(x) for x in X]

# *********** Aquí termina la sección de agregar código *************

In [14]:
# Despleguemos los primeros comentarios después de tu proceso de limpieza:

for x in Xclean[0:5]:
  print(x)


['so', 'there', 'is', 'no', 'way', 'for', 'me', 'to', 'plug', 'it', 'in', 'here', 'in', 'the', 'us', 'unless', 'go', 'by', 'converter']
['good', 'case', 'excellent', 'value']
['great', 'for', 'the', 'jawbone']
['tied', 'to', 'charger', 'for', 'conversations', 'lasting', 'more', 'than', 'minutesmajor', 'problems']
['the', 'mic', 'is', 'great']


# **Pregunta - 3:**



Realicemos una partición aleatoria con los porcentajes que consideres más adecuados. Utiliza una semilla para su reproducibilidad.

In [15]:

# ************* Inicia la sección de agregar código:*****************************

Xtrain, x_val_and_test, Ytrain, y_val_and_test = train_test_split(Xclean, y, train_size=.70, shuffle=True, random_state=1)
Xval, Xtest, Yval, Ytest = train_test_split(x_val_and_test, y_val_and_test, test_size=.50, shuffle=True, random_state=1)

In [16]:
print('X,y Train:', len(Xtrain), len(Ytrain))
print('X,y Val:', len(Xval), len(Yval))
print('X,y Test', len(Xtest), len(Ytest))

# *********** Termina la sección de agregar código *************

X,y Train: 2100 2100
X,y Val: 450 450
X,y Test 450 450


# **Pregunta - 4:**




### **Construye tu vocabulario a continuación utilizando solamente el conjunto de Train:**


In [17]:
# a.	Usa el conjunto de entrenamiento para generar tu vocabulario
#     con un tamaño que consideres adecuado:


# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********

token_dict = Counter()

for k in range(len(Xtrain)):
  token_dict.update(Xtrain[k])


print('Longitud del diccionario:', len(token_dict))
print('\n(word,frequency):')
print(token_dict.most_common(10))


Longitud del diccionario: 4216

(word,frequency):
[('the', 1350), ('and', 753), ('is', 536), ('it', 485), ('to', 472), ('this', 449), ('of', 423), ('was', 413), ('in', 280), ('for', 242)]


In [18]:

min_freq = 5

midicc = {}

for token, freq in token_dict.items():

    if freq >= min_freq:
        midicc[token] = freq

# *********** Aquí termina la sección de agregar código *************

In [19]:
# b.	Indica el tamaño del vocabulario generado.

# ******* Inicia la sección de agregar código: ***********

print('Longitud del vocabulario generado:', len(midicc))

# *********** Aquí termina la sección de agregar código *************

Longitud del vocabulario generado: 647


In [20]:
# c.	Con el vocabulario generado, filtra los conjuntos de entrenamiento,
#     validación y prueba para que todos los comentarios usen solamente las
#     palabras de este vocabulario.

#     Llamar train_X, val_X y test_X a estos tres conjuntos.


# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


train_X = []
for ss in Xtrain:
  train_X.append([w for w in ss if w in midicc])

val_X = []
for ss in Xval:
  val_X.append([w for w in ss if w in midicc])

test_X = []
for ss in Xtest:
  test_X.append([w for w in ss if w in midicc])


# *********** Aquí termina la sección de agregar código *************


In [21]:
# Vemos el resultado de los primeros comentarios del conjunto de validación:

for ss in val_X[0:5]:
  print(ss)

['director', 'to', 'from', 'and', 'with', 'the', 'drama', 'and', 'comedy', 'you', 'can', 'never', 'tell', 'what', 'is', 'going', 'to', 'next']
['food', 'is', 'way', 'overpriced', 'and', 'are', 'small']
['and', 'good', 'prices']
['this', 'place', 'is', 'two', 'up']
['its', 'fun', 'fast', 'and', 'of', 'the', 'night', 'of']


# **Pregunta - 5:**

Incluye tus comentarios sobre cada modelo de HuggingFace indicado.

### ++++++++ Inicia la sección de agregar texto: +++++++++++

* **a) bge-base-en-v1.5**
- Es un modelo de embeddings desarrollado por Beijing Academy of Artificial Intelligence (BAAI).
- Soporta tareas de búsqueda semántica (retrieval), similitud textual y sistemas RAG.
- Fue entrenado principalmente para textos en inglés.
- Cuenta con 12 capas (layers).
- El tamaño del vector embebido es de 768 dimensiones.
- Tiene un buen equilibrio entre rendimiento, velocidad y consumo de memoria.


* **b) bge-large-en-v1.5**
- Es una versión más grande y precisa de la familia BGE.
- Está optimizado para tareas de recuperación de información y búsqueda semántica.
- Cuenta con 24 capas (layers).
- El tamaño del embedding es de 1024 dimensiones.
- Genera representaciones semánticas más precisas que la versión base.
- Requiere más memoria RAM/VRAM y mayor tiempo de procesamiento.

* **c)	e5-base-v2**
- Es un modelo de embeddings desarrollado por Microsoft.
- Está diseñado para tareas de recuperación semántica y comparación de textos.
- Cuenta con 12 capas (layers).
El tamaño del vector embebido es de 768 dimensiones.
- Funciona principalmente con textos en inglés.
- Los textos largos se truncan a un máximo de 512 tokens

### ++++++++ Termina la sección de agregar texto: +++++++++++

# **Pregunta - 6:**

In [22]:
# a) Cargar el modelo de embeddings de HuggingFace seleccionado:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


from sentence_transformers import SentenceTransformer

modelo_emb = SentenceTransformer('BAAI/bge-base-en-v1.5')

palabras_vocab = list(midicc.keys())
vectores_palabras = modelo_emb.encode(palabras_vocab, show_progress_bar=False)

diccionario_emb = {palabra: vector for palabra, vector in zip(palabras_vocab, vectores_palabras)}


# *********** Aquí termina la sección de agregar código *************

C:\Users\lfnalvaradof\AppData\Local\Programs\Python\Python312\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm



Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]


Loading weights: 100%|██████████| 199/199 [00:00<00:00, 7960.12it/s]

In [23]:
# b) Primeros 3 elementos clave:valor del diccionario generado.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


for palabra, vector in list(diccionario_emb.items())[:3]:
    print(f'{palabra}: {vector[:5]}...')


# *********** Aquí termina la sección de agregar código *************



his: [-0.01901003  0.03201386  0.02433447 -0.00354911  0.01592904]...
dont: [0.00957289 0.03321743 0.02284432 0.01627553 0.03065626]...
much: [-0.04203335 -0.020161    0.0199551  -0.03272091 -0.00774411]...


In [24]:
# c) Tamaño del diccionario generado:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


print('Tamaño del diccionario:', len(diccionario_emb))


# *********** Aquí termina la sección de agregar código *************


Tamaño del diccionario: 647


# **Pregunta - 7:**




Generamos los vectores embebidos a partir de los conjuntos de entrenamiento, validación y prueba y con las características indicadas en el archivo PDF.

Los llamaremos trainEmb, valEmb y testEmb, respectivamente.


In [25]:
# a) Comentarios con vectores embebidos.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


def doc_a_vector(doc, dic):
    vectores = [dic[w] for w in doc if w in dic]
    if vectores:
        return np.mean(vectores, axis=0)
    return np.zeros(len(next(iter(dic.values()))))

trainEmb = np.array([doc_a_vector(doc, diccionario_emb) for doc in train_X])
valEmb   = np.array([doc_a_vector(doc, diccionario_emb) for doc in val_X])
testEmb  = np.array([doc_a_vector(doc, diccionario_emb) for doc in test_X])


# *********** Aquí termina la sección de agregar código *************

In [26]:
# b) Dimensiones de los conjuntos trainEmb, valEmb y testEmb.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


print('trainEmb:', trainEmb.shape)
print('valEmb:  ', valEmb.shape)
print('testEmb: ', testEmb.shape)


# *********** Aquí termina la sección de agregar código *************

trainEmb: (2100, 768)
valEmb:   (450, 768)
testEmb:  (450, 768)


# **Pregunta - 8:**

In [27]:
# Número de tokens generedos al obtener cada uno de los conjuntos trainEmb, valEmb y testEmb.

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


tokens_train = sum(len([w for w in doc if w in diccionario_emb]) for doc in train_X)
tokens_val   = sum(len([w for w in doc if w in diccionario_emb]) for doc in val_X)
tokens_test  = sum(len([w for w in doc if w in diccionario_emb]) for doc in test_X)

print('Tokens en trainEmb:', tokens_train)
print('Tokens en valEmb:  ', tokens_val)
print('Tokens en testEmb: ', tokens_test)


# *********** Aquí termina la sección de agregar código *************

Tokens en trainEmb: 18179
Tokens en valEmb:   3874
Tokens en testEmb:  3544


# **Pregunta - 9:**



Entrenamiento y reporte de los modelos de Regresión Logística y Bosque Aleatorio (Random Forest).


In [28]:
# 9a) REGRESIÓN LOGÍSTICA:

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

modeloLR = LogisticRegression(C=1.0, max_iter=1000, random_state=1)
modeloLR.fit(trainEmb, Ytrain)

acc_train_LR = accuracy_score(Ytrain, modeloLR.predict(trainEmb))
acc_val_LR   = accuracy_score(Yval,   modeloLR.predict(valEmb))

print(f'LR - Exactitud Train: {acc_train_LR:.4f}')
print(f'LR - Exactitud Val:   {acc_val_LR:.4f}')
print()
print(classification_report(Yval, modeloLR.predict(valEmb), target_names=['Negativo', 'Positivo']))


# *********** Aquí termina la sección de agregar código *************


LR - Exactitud Train: 0.8081
LR - Exactitud Val:   0.7689

              precision    recall  f1-score   support

    Negativo       0.75      0.79      0.77       218
    Positivo       0.79      0.75      0.77       232

    accuracy                           0.77       450
   macro avg       0.77      0.77      0.77       450
weighted avg       0.77      0.77      0.77       450



In [29]:
# 9b) BOSQUE ALEATORIO (Random Forest):

# ******* Inlcuye a continuación todas las líneas de código y celdas que requieras: ***********


from sklearn.ensemble import RandomForestClassifier

modeloRF = RandomForestClassifier(n_estimators=200, max_depth=5, min_samples_leaf=5, random_state=1)
modeloRF.fit(trainEmb, Ytrain)

acc_train_RF = accuracy_score(Ytrain, modeloRF.predict(trainEmb))
acc_val_RF   = accuracy_score(Yval,   modeloRF.predict(valEmb))

print(f'RF - Exactitud Train: {acc_train_RF:.4f}')
print(f'RF - Exactitud Val:   {acc_val_RF:.4f}')
print()
print(classification_report(Yval, modeloRF.predict(valEmb), target_names=['Negativo', 'Positivo']))


# *********** Aquí termina la sección de agregar código *************

RF - Exactitud Train: 0.9081
RF - Exactitud Val:   0.7422

              precision    recall  f1-score   support

    Negativo       0.72      0.77      0.74       218
    Positivo       0.77      0.72      0.74       232

    accuracy                           0.74       450
   macro avg       0.74      0.74      0.74       450
weighted avg       0.74      0.74      0.74       450



# **Pregunta - 10**

**Proceso basado en modelos Preentrenados**

In [30]:
# 10a) Particion.:

# ******* Incluye a continuacion todas las lineas de codigo y celdas que requieras: ***********


Xtrain_pt, x_val_and_test_pt, Ytrain_pt, y_val_and_test_pt = train_test_split(
    df.review.astype(str),
    df.label,
    train_size=.70,
    shuffle=True,
    random_state=1,
    stratify=df.label
)

Xval_pt, Xtest_pt, Yval_pt, Ytest_pt = train_test_split(
    x_val_and_test_pt,
    y_val_and_test_pt,
    test_size=.50,
    shuffle=True,
    random_state=1,
    stratify=y_val_and_test_pt
)

print('X,y Train:', len(Xtrain_pt), len(Ytrain_pt))
print('X,y Val:  ', len(Xval_pt), len(Yval_pt))
print('X,y Test: ', len(Xtest_pt), len(Ytest_pt))
print('Distribucion Train:', Ytrain_pt.value_counts().sort_index().to_dict())
print('Distribucion Val:  ', Yval_pt.value_counts().sort_index().to_dict())
print('Distribucion Test: ', Ytest_pt.value_counts().sort_index().to_dict())


# *********** Aqui termina la seccion de agregar codigo *************

X,y Train: 2100 2100
X,y Val:   450 450
X,y Test:  450 450
Distribucion Train: {0: 1050, 1: 1050}
Distribucion Val:   {0: 225, 1: 225}
Distribucion Test:  {0: 225, 1: 225}


In [31]:
# 10b) Vectores embebidos:

# ******* Incluye a continuacion todas las lineas de codigo y celdas que requieras: ***********


# En este enfoque se usa directamente cada resena completa como entrada del
# modelo preentrenado, sin promediar embeddings de palabras del vocabulario.
modelo_emb_pt = modelo_emb if 'modelo_emb' in globals() else SentenceTransformer('BAAI/bge-base-en-v1.5')

trainEmb_pt = modelo_emb_pt.encode(
    Xtrain_pt.tolist(), batch_size=32, show_progress_bar=True, normalize_embeddings=True
)
valEmb_pt = modelo_emb_pt.encode(
    Xval_pt.tolist(), batch_size=32, show_progress_bar=True, normalize_embeddings=True
)
testEmb_pt = modelo_emb_pt.encode(
    Xtest_pt.tolist(), batch_size=32, show_progress_bar=True, normalize_embeddings=True
)

print('trainEmb_pt:', trainEmb_pt.shape)
print('valEmb_pt:  ', valEmb_pt.shape)
print('testEmb_pt: ', testEmb_pt.shape)


# *********** Aqui termina la seccion de agregar codigo *************


Batches:   0%|          | 0/66 [00:00<?, ?it/s]


Batches:   2%|▏         | 1/66 [00:01<01:59,  1.84s/it]


Batches:   3%|▎         | 2/66 [00:02<01:28,  1.39s/it]


Batches:   5%|▍         | 3/66 [00:03<01:12,  1.16s/it]


Batches:   6%|▌         | 4/66 [00:04<01:05,  1.06s/it]


Batches:   8%|▊         | 5/66 [00:05<01:00,  1.01it/s]


Batches:   9%|▉         | 6/66 [00:06<00:54,  1.10it/s]


Batches:  11%|█         | 7/66 [00:07<00:50,  1.16it/s]


Batches:  12%|█▏        | 8/66 [00:07<00:48,  1.20it/s]


Batches:  14%|█▎        | 9/66 [00:08<00:47,  1.21it/s]


Batches:  15%|█▌        | 10/66 [00:09<00:44,  1.26it/s]


Batches:  17%|█▋        | 11/66 [00:10<00:42,  1.29it/s]


Batches:  18%|█▊        | 12/66 [00:10<00:39,  1.38it/s]


Batches:  20%|█▉        | 13/66 [00:11<00:39,  1.34it/s]


Batches:  21%|██        | 14/66 [00:12<00:37,  1.38it/s]


Batches:  23%|██▎       | 15/66 [00:12<00:36,  1.40it/s]


Batches:  24%|██▍       | 16/66 [00:13<00:35,  1.43it/s]


Batches:  26%|██▌       | 17/66 [00:14<00:35,  1.36it/s]


Batches:  27%|██▋       | 18/66 [00:15<00:34,  1.40it/s]


Batches:  29%|██▉       | 19/66 [00:15<00:32,  1.45it/s]


Batches:  30%|███       | 20/66 [00:16<00:31,  1.46it/s]


Batches:  32%|███▏      | 21/66 [00:16<00:28,  1.55it/s]


Batches:  33%|███▎      | 22/66 [00:17<00:27,  1.61it/s]


Batches:  35%|███▍      | 23/66 [00:17<00:25,  1.72it/s]


Batches:  36%|███▋      | 24/66 [00:18<00:23,  1.77it/s]


Batches:  38%|███▊      | 25/66 [00:19<00:22,  1.80it/s]


Batches:  39%|███▉      | 26/66 [00:19<00:22,  1.75it/s]


Batches:  41%|████      | 27/66 [00:20<00:22,  1.74it/s]


Batches:  42%|████▏     | 28/66 [00:20<00:21,  1.80it/s]


Batches:  44%|████▍     | 29/66 [00:21<00:19,  1.87it/s]


Batches:  45%|████▌     | 30/66 [00:21<00:18,  1.99it/s]


Batches:  47%|████▋     | 31/66 [00:22<00:16,  2.07it/s]


Batches:  48%|████▊     | 32/66 [00:22<00:16,  2.00it/s]


Batches:  50%|█████     | 33/66 [00:23<00:18,  1.80it/s]


Batches:  52%|█████▏    | 34/66 [00:23<00:16,  1.92it/s]


Batches:  53%|█████▎    | 35/66 [00:24<00:15,  1.99it/s]


Batches:  55%|█████▍    | 36/66 [00:24<00:14,  2.07it/s]


Batches:  56%|█████▌    | 37/66 [00:25<00:13,  2.13it/s]


Batches:  58%|█████▊    | 38/66 [00:25<00:12,  2.29it/s]


Batches:  59%|█████▉    | 39/66 [00:25<00:11,  2.33it/s]


Batches:  61%|██████    | 40/66 [00:26<00:10,  2.44it/s]


Batches:  62%|██████▏   | 41/66 [00:26<00:10,  2.33it/s]


Batches:  64%|██████▎   | 42/66 [00:27<00:09,  2.43it/s]


Batches:  65%|██████▌   | 43/66 [00:27<00:08,  2.60it/s]


Batches:  67%|██████▋   | 44/66 [00:27<00:08,  2.60it/s]


Batches:  68%|██████▊   | 45/66 [00:28<00:07,  2.70it/s]


Batches:  70%|██████▉   | 46/66 [00:28<00:07,  2.76it/s]


Batches:  71%|███████   | 47/66 [00:28<00:06,  2.91it/s]


Batches:  73%|███████▎  | 48/66 [00:29<00:06,  2.97it/s]


Batches:  74%|███████▍  | 49/66 [00:29<00:06,  2.68it/s]


Batches:  76%|███████▌  | 50/66 [00:29<00:05,  2.76it/s]


Batches:  77%|███████▋  | 51/66 [00:30<00:05,  2.84it/s]


Batches:  79%|███████▉  | 52/66 [00:30<00:04,  3.00it/s]


Batches:  80%|████████  | 53/66 [00:30<00:04,  2.96it/s]


Batches:  82%|████████▏ | 54/66 [00:31<00:03,  3.14it/s]


Batches:  83%|████████▎ | 55/66 [00:31<00:03,  3.19it/s]


Batches:  85%|████████▍ | 56/66 [00:31<00:03,  3.32it/s]


Batches:  86%|████████▋ | 57/66 [00:31<00:02,  3.48it/s]


Batches:  88%|████████▊ | 58/66 [00:32<00:02,  3.58it/s]


Batches:  89%|████████▉ | 59/66 [00:32<00:01,  3.74it/s]


Batches:  91%|█████████ | 60/66 [00:32<00:01,  3.80it/s]


Batches:  92%|█████████▏| 61/66 [00:32<00:01,  3.92it/s]


Batches:  94%|█████████▍| 62/66 [00:33<00:01,  3.90it/s]


Batches:  95%|█████████▌| 63/66 [00:33<00:00,  3.98it/s]


Batches:  97%|█████████▋| 64/66 [00:33<00:00,  3.90it/s]


Batches:  98%|█████████▊| 65/66 [00:33<00:00,  4.15it/s]


Batches: 100%|██████████| 66/66 [00:34<00:00,  4.63it/s]


Batches: 100%|██████████| 66/66 [00:34<00:00,  1.94it/s]


Batches:   0%|          | 0/15 [00:00<?, ?it/s]


Batches:   7%|▋         | 1/15 [00:01<00:14,  1.07s/it]


Batches:  13%|█▎        | 2/15 [00:01<00:11,  1.15it/s]


Batches:  20%|██        | 3/15 [00:02<00:09,  1.25it/s]


Batches:  27%|██▋       | 4/15 [00:03<00:07,  1.38it/s]


Batches:  33%|███▎      | 5/15 [00:03<00:06,  1.50it/s]


Batches:  40%|████      | 6/15 [00:04<00:05,  1.66it/s]


Batches:  47%|████▋     | 7/15 [00:04<00:04,  1.73it/s]


Batches:  53%|█████▎    | 8/15 [00:05<00:03,  1.78it/s]


Batches:  60%|██████    | 9/15 [00:05<00:03,  1.99it/s]


Batches:  67%|██████▋   | 10/15 [00:05<00:02,  2.20it/s]


Batches:  73%|███████▎  | 11/15 [00:06<00:01,  2.48it/s]


Batches:  80%|████████  | 12/15 [00:06<00:01,  2.59it/s]


Batches:  87%|████████▋ | 13/15 [00:06<00:00,  2.90it/s]


Batches:  93%|█████████▎| 14/15 [00:07<00:00,  3.26it/s]


Batches: 100%|██████████| 15/15 [00:07<00:00,  2.12it/s]


Batches:   0%|          | 0/15 [00:00<?, ?it/s]


Batches:   7%|▋         | 1/15 [00:02<00:29,  2.08s/it]


Batches:  13%|█▎        | 2/15 [00:02<00:16,  1.30s/it]


Batches:  20%|██        | 3/15 [00:03<00:12,  1.03s/it]


Batches:  27%|██▋       | 4/15 [00:04<00:09,  1.16it/s]


Batches:  33%|███▎      | 5/15 [00:04<00:07,  1.30it/s]


Batches:  40%|████      | 6/15 [00:05<00:06,  1.46it/s]


Batches:  47%|████▋     | 7/15 [00:05<00:04,  1.66it/s]


Batches:  53%|█████▎    | 8/15 [00:06<00:03,  1.84it/s]


Batches:  60%|██████    | 9/15 [00:06<00:03,  1.98it/s]


Batches:  67%|██████▋   | 10/15 [00:06<00:02,  2.22it/s]


Batches:  73%|███████▎  | 11/15 [00:07<00:01,  2.32it/s]


Batches:  80%|████████  | 12/15 [00:07<00:01,  2.59it/s]


Batches:  87%|████████▋ | 13/15 [00:07<00:00,  2.85it/s]


Batches:  93%|█████████▎| 14/15 [00:08<00:00,  3.07it/s]


Batches: 100%|██████████| 15/15 [00:08<00:00,  1.85it/s]

trainEmb_pt: (2100, 768)
valEmb_pt:   (450, 768)
testEmb_pt:  (450, 768)


In [32]:
# 10c) REGRESION LOGISTICA.

# ******* Incluye a continuacion todas las lineas de codigo y celdas que requieras: ***********


modeloLR_pt = LogisticRegression(C=1.0, max_iter=1000, random_state=1)
modeloLR_pt.fit(trainEmb_pt, Ytrain_pt)

acc_train_LR_pt = accuracy_score(Ytrain_pt, modeloLR_pt.predict(trainEmb_pt))
acc_val_LR_pt = accuracy_score(Yval_pt, modeloLR_pt.predict(valEmb_pt))

print(f'LR PT - Exactitud Train: {acc_train_LR_pt:.4f}')
print(f'LR PT - Exactitud Val:   {acc_val_LR_pt:.4f}')
print()
print(classification_report(Yval_pt, modeloLR_pt.predict(valEmb_pt), target_names=['Negativo', 'Positivo']))


# *********** Aqui termina la seccion de agregar codigo *************

LR PT - Exactitud Train: 0.9671
LR PT - Exactitud Val:   0.9733

              precision    recall  f1-score   support

    Negativo       0.97      0.97      0.97       225
    Positivo       0.97      0.97      0.97       225

    accuracy                           0.97       450
   macro avg       0.97      0.97      0.97       450
weighted avg       0.97      0.97      0.97       450



In [33]:
# 10d) BOSQUE ALEATORIO.

# ******* Incluye a continuacion todas las lineas de codigo y celdas que requieras: ***********


modeloRF_pt = RandomForestClassifier(
    n_estimators=200,
    max_depth=8,
    min_samples_leaf=3,
    random_state=1,
    n_jobs=-1
)
modeloRF_pt.fit(trainEmb_pt, Ytrain_pt)

acc_train_RF_pt = accuracy_score(Ytrain_pt, modeloRF_pt.predict(trainEmb_pt))
acc_val_RF_pt = accuracy_score(Yval_pt, modeloRF_pt.predict(valEmb_pt))

print(f'RF PT - Exactitud Train: {acc_train_RF_pt:.4f}')
print(f'RF PT - Exactitud Val:   {acc_val_RF_pt:.4f}')
print()
print(classification_report(Yval_pt, modeloRF_pt.predict(valEmb_pt), target_names=['Negativo', 'Positivo']))


# *********** Aqui termina la seccion de agregar codigo *************

RF PT - Exactitud Train: 0.9905
RF PT - Exactitud Val:   0.9711

              precision    recall  f1-score   support

    Negativo       0.98      0.96      0.97       225
    Positivo       0.96      0.98      0.97       225

    accuracy                           0.97       450
   macro avg       0.97      0.97      0.97       450
weighted avg       0.97      0.97      0.97       450



# **Pregunta - 11:**

In [34]:
# Reporte del mejor modelo y particion con el conjunto de Prueba.

# ******* Incluye a continuacion todas las lineas de codigo y celdas que requieras: ***********


resultados_validacion = pd.DataFrame([
    {'Representacion': 'Promedio de embeddings por vocabulario', 'Modelo': 'Regresion Logistica', 'Train': acc_train_LR, 'Validacion': acc_val_LR},
    {'Representacion': 'Promedio de embeddings por vocabulario', 'Modelo': 'Bosque Aleatorio', 'Train': acc_train_RF, 'Validacion': acc_val_RF},
    {'Representacion': 'Embeddings preentrenados directos', 'Modelo': 'Regresion Logistica', 'Train': acc_train_LR_pt, 'Validacion': acc_val_LR_pt},
    {'Representacion': 'Embeddings preentrenados directos', 'Modelo': 'Bosque Aleatorio', 'Train': acc_train_RF_pt, 'Validacion': acc_val_RF_pt},
])

print('Comparacion en validacion:')
print(resultados_validacion.sort_values('Validacion', ascending=False).to_string(index=False))

mejor_fila = resultados_validacion.sort_values('Validacion', ascending=False).iloc[0]
print()
print('Mejor modelo seleccionado por validacion:')
print(f"{mejor_fila['Modelo']} con {mejor_fila['Representacion']}")

if mejor_fila['Representacion'] == 'Embeddings preentrenados directos':
    mejor_modelo = modeloLR_pt if mejor_fila['Modelo'] == 'Regresion Logistica' else modeloRF_pt
    mejor_testEmb = testEmb_pt
    mejor_Ytest = Ytest_pt
else:
    mejor_modelo = modeloLR if mejor_fila['Modelo'] == 'Regresion Logistica' else modeloRF
    mejor_testEmb = testEmb
    mejor_Ytest = Ytest

y_pred_test_mejor = mejor_modelo.predict(mejor_testEmb)
acc_test_mejor = accuracy_score(mejor_Ytest, y_pred_test_mejor)

print()
print(f'Exactitud del mejor modelo en Prueba: {acc_test_mejor:.4f}')
print()
print(classification_report(mejor_Ytest, y_pred_test_mejor, target_names=['Negativo', 'Positivo']))


# *********** Aqui termina la seccion de agregar codigo *************

Comparacion en validacion:
                        Representacion              Modelo    Train  Validacion
     Embeddings preentrenados directos Regresion Logistica 0.967143    0.973333
     Embeddings preentrenados directos    Bosque Aleatorio 0.990476    0.971111
Promedio de embeddings por vocabulario Regresion Logistica 0.808095    0.768889
Promedio de embeddings por vocabulario    Bosque Aleatorio 0.908095    0.742222

Mejor modelo seleccionado por validacion:
Regresion Logistica con Embeddings preentrenados directos

Exactitud del mejor modelo en Prueba: 0.9689

              precision    recall  f1-score   support

    Negativo       0.98      0.96      0.97       225
    Positivo       0.96      0.98      0.97       225

    accuracy                           0.97       450
   macro avg       0.97      0.97      0.97       450
weighted avg       0.97      0.97      0.97       450



# **Pregunta - 12:**


Incluye tus comentarios finales de la actividad.

### ++++++++ Inicia la seccion de agregar texto: +++++++++++


### Preprocesamiento y vocabulario

- Se aplico limpieza basica: conversion a minusculas, eliminacion de puntuacion y digitos, tokenizacion con expresiones regulares y descarte de tokens de un solo caracter.
- Para construir el vocabulario se uso unicamente el conjunto de entrenamiento, con un umbral de **min_freq = 5**. Esto resulto en un vocabulario de **647 palabras**, eliminando terminos muy raros que podrian introducir ruido sin aportar capacidad de generalizacion.
- La primera particion fue **70% entrenamiento / 15% validacion / 15% prueba** con `random_state=1`, resultando en 2100, 450 y 450 muestras respectivamente.

### Enfoques de embeddings evaluados

- En las preguntas 6 a 9 se genero un diccionario `{palabra: vector}` con **bge-base-en-v1.5** y cada resena se represento como el promedio de los embeddings de sus tokens presentes en el vocabulario.
- En la pregunta 10 se uso un segundo enfoque basado directamente en modelos preentrenados: cada resena completa se envio a **bge-base-en-v1.5** y se obtuvo un embedding semantico de 768 dimensiones para el documento completo.
- Para el enfoque preentrenado directo se utilizo una particion estratificada 70/15/15, manteniendo 1050 resenas negativas y 1050 positivas en entrenamiento, y 225/225 en validacion y prueba.

### Resultados principales

| Representacion | Modelo | Exactitud Train | Exactitud Val |
|---|---:|---:|---:|
| Promedio de embeddings por vocabulario | Regresion Logistica | 80.81% | 76.89% |
| Promedio de embeddings por vocabulario | Bosque Aleatorio | 90.81% | 74.22% |
| Embeddings preentrenados directos | Regresion Logistica | 96.71% | 97.33% |
| Embeddings preentrenados directos | Bosque Aleatorio | 99.05% | 97.11% |

- El enfoque de embeddings preentrenados directos supera ampliamente al promedio de embeddings por palabra, ya que conserva mejor el contexto semantico de la resena completa.
- La **Regresion Logistica con embeddings preentrenados directos** fue el mejor modelo por validacion, con **97.33%** de exactitud.
- Al evaluarlo en el conjunto de prueba, el mejor modelo obtuvo **96.89%** de exactitud, con precision, recall y F1 cercanos a **0.97** en ambas clases.
- Random Forest tambien tuvo muy buen desempeno con embeddings directos, pero mostro una brecha train-validacion mayor que Regresion Logistica, por lo que la Regresion Logistica ofrece una solucion mas simple y con mejor generalizacion.

### Conclusion

Los embeddings preentrenados de Hugging Face permiten representar mejor el significado de comentarios cortos que una agregacion promedio de vectores por palabra. Para este conjunto de datos balanceado, una Regresion Logistica entrenada sobre embeddings directos de **bge-base-en-v1.5** fue suficiente para obtener un desempeno alto, estable y equilibrado entre clases positivas y negativas.


### ++++++++ Termina la seccion de agregar texto: +++++++++++

# **Fin de la Actividad de Vectores Embebidos - HuggingFace**